### PR TITLE
add eldoc support

### PR DESCRIPTION
**Pull request by otijhuis**
This probably needs some cleaning up and maybe some changes.
Would be really nice if you could take a look.

I started out with the code from cider and changed things where needed. Haven't hit any performance issues so far. Eldoc works in clojure buffer and repl.

### CODE REVIEW

**Comment by bbatsov** [comment]
I'd suggest killing the separate files and just putting everything in `inf-clojure.el`. The package is way simpler than CIDER, so breaking it up doesn't add much value. I'd also suggest against the backports from `subr-x`, as the codebase is tiny and we don't really need them that much.

**Comment by bbatsov** [comment]
P.S. Capitalize the commit messages.

**Comment by otijhuis** [comment]
Sure thing. I just kept the same structure for now so it's easier for me to find things in Cider.
Thanks for the feedback :). I'll fix those things soon.

Currently trying to get it working for ClojureScript, which is a bit of a pain it seems.

**Comment by bbatsov** [comment]
OK, take your time. We're definitely not in a hurry.

**Comment by otijhuis** [comment]
I cleaned things up and added everything that was needed to `inf-clojure.el`. Not enabling eldoc mode when the major mode is ClojureScript currently. Still looking for a solution. Asked Bruce Hauman for some help as well but he didn't see an easy solution either.

Because of ClojureScript I haven't enabled eldoc mode in the REPL by default.

You can run
```elisp
(inf-clojure-eldoc-setup)
(eldoc-mode +1)
```
when the `inf-clojure` buffer is active and then it works fine in the REPL as well. For Clojure that is.

**Comment by bbatsov** [comment]
Guess you can add this info the README, so it'd be easier for users to find it.

**Comment by otijhuis** [comment]
Sure thing. I'll add it. Or is there an easy way to find out if a cljs or clj REPL is running?

**Comment by otijhuis** [comment]
Just noticed that I can always do the `(inf-clojure-eldoc-setup)` anyway. Saves a step. I'll update the PR.

**Comment by otijhuis** [comment]
Only `M-x eldoc-mode` needed now in `*inf-clojure*` buffer. Updated README.

File: inf-clojure.el [hunk 1]
```diff
@@ -131,6 +134,10 @@ The following commands are available:
 
 \\{inf-clojure-minor-mode-map}"
   :lighter "" :keymap inf-clojure-minor-mode-map
+  (inf-clojure-eldoc-setup)
+  ;; No ElDoc support for ClojureScript yet
+  (unless (equal major-mode 'clojurescript-mode)
```
Comment (bbatsov) [inline inf-clojure.el]: It's generally better to use `derived-mode-p`, as it will work with derived modes as well.

**Comment by bbatsov** [comment]
Btw, I now remembered we don't turn on `eldoc-mode` by default in CIDER, so we should probably not turn it on by default here as well. We just advise people there to manually add it to their mode hooks.

Btw, I'm guessing the eldoc-code could check that's in a Clojure REPL, so that's it's safe to have it turned on in a ClojureScript REPL.

**Comment by otijhuis** [comment]
Ok, no problem. I'll change it to off by default then.

Been thinking about doing the check in the eldoc-code but not sure what the best, and performant, way is to check if it's a Clojure REPL.

**Comment by bbatsov** [comment]
Guess you can simply check about the existence of some cljs ns or dynamic var. That should be quite fast.

**Comment by otijhuis** [comment]
Updated the PR:

- eldoc is no longer enabled by default
- deleted a debug message I forgot to remove
- updated the README on how to enable eldoc

In ClojureScript the debug message was still showing before, this made it annoying.
Tried it just now with ClojureScript. No problems, just won't show anything in the echo area.
So you should be able to leave ElDoc enabled, even in cljs buffers/REPLs. Haven't seen
any troublesome performance issues so far.

**Comment by bbatsov** [comment]
Apart from my small remarks, things are looking good. We're pretty close to merging this.

**Comment by otijhuis** [comment]
Sorry about the `cl-lib`. Copied a lot from `cider-eldoc.el` but turned out I didn't need certain functions. Those used `cl-lib`.

**Comment by otijhuis** [comment]
Already removed it.

**Comment by otijhuis** [comment]
Did you have any other remarks besides `cl-lib`? Force pushed some stuff to keep it clean. Guess that removed the comments as well.

Btw, what is your regular workflow with PR's? Do I need to force push or fix things in new commits?
Still figuring out the best way to do things. If you have any advice let me know. Want to make it as easy for you as possible and I'm still rather new at this PR stuff :)

**Comment by bbatsov** [comment]
> Did you have any other remarks besides cl-lib? Force pushed some stuff to keep it clean. Guess that removed the comments as well.

They were about the text in the README. Guess you can see the email notification or check github's notifications centre. 

> Btw, what is your regular workflow with PR's? Do I need to force push or fix things in new commits?
> Still figuring out the best way to do things. If you have any advice let me know. Want to make it as easy for you as possible and I'm still rather new at this PR stuff :)

I also address comments by force-pushing. Doing new commits is fine as well, as long as they get properly squashed before being merged in the end.

**Comment by otijhuis** [comment]
Ok, thanks for the info. Updated the README.

**Pull request merged**